### PR TITLE
fix(connection): getConnectionCredentials should return credentials even on error

### PR DESCRIPTION
**Pull request by bodinsamuel**
## Describe your changes

Fixes https://linear.app/nango/issue/NAN-1933/fix-ui-crash-when-loading-a-connection-with-token-refresh-error

- Split getConnectionCredentials and initial connection fetch
For the /connections/:id endpoint we still want to return the connection even if the refresh failed but if there was an error we stripped the credentials to avoid leaking stuff. Initially, I had just reverted this and stripped the credentials outside of the function at multiple places but the code was even more crappier.
So I split the function, this allows us to not repeat call to fetch the integration on the API and correctly output the errors. It's a bit less straightforward but more granular

### CODE REVIEW

File: packages/server/lib/refreshConnections.ts [hunk 1]
```diff
@@ -56,21 +56,22 @@ export async function exec(): Promise<void> {
             const limit = 1000;
             // eslint-disable-next-line no-constant-condition
             while (true) {
-                const staleConnections = await connectionService.getStaleConnections({ days: 0, limit, cursor });
+                const staleConnections = await connectionService.getStaleConnections({ days: 1, limit, cursor });
```
Comment (bodinsamuel) [inline packages/server/lib/refreshConnections.ts:59]: I think it was an error, let me know @TBonnin

File: packages/server/lib/hooks/connection/post-connection.ts [hunk 2]
```diff
@@ -23,22 +22,12 @@ async function execute(createdConnection: RecentlyCreatedConnection, providerNam
...
-            providerConfigKey: upsertedConnection.provider_config_key,
-            logContextGetter,
-            instantRefresh: false,
-            onRefreshSuccess: connectionRefreshSuccessHook,
-            onRefreshFailed: connectionRefreshFailedHook
-        });
-
-        if (credentialResponse.isErr()) {
+        const connectionRes = await connectionService.getConnection(upsertedConnection.connection_id, upsertedConnection.provider_config_key, environment.id);
+        if (connectionRes.error || !connectionRes.response) {
```
Comment (bodinsamuel) [inline packages/server/lib/hooks/connection/post-connection.ts:26]: On post connection I don't think we need to test again the credentials since we just fetched them, but maybe I forgot an edge case? @khaliqgant

File: packages/server/lib/refreshConnections.ts (hunk 1 above)
Comment (TBonnin) [reply packages/server/lib/refreshConnections.ts:59]: yes, that's a test leftover. Thank you for fixing

File: packages/server/lib/refreshConnections.ts [hunk 3]
```diff
@@ -36,7 +36,7 @@ export function refreshConnectionsCron(): void {
 }
 
 export async function exec(): Promise<void> {
-    return await tracer.trace<Promise<void>>('nango.server.cron.refreshConnections', async (span) => {
+    await tracer.trace<Promise<void>>('nango.server.cron.refreshConnections', async (span) => {
```
Comment (TBonnin) [inline packages/server/lib/refreshConnections.ts:39]: exec is itself awaited above.

File: packages/shared/lib/services/connection.service.ts [hunk 4]
```diff
@@ -985,55 +985,26 @@ class ConnectionService {
...
-        }
-
-        const provider = getProvider(config?.provider);
+        const provider = getProvider(integration.provider);
         if (!provider) {
             const error = new NangoError('unknown_provider_config');
             return Err(error);
         }
 
+        const copy = { ...connection };
```
Comment (TBonnin) [inline packages/shared/lib/services/connection.service.ts:994]: why a copy?

File: packages/server/lib/controllers/v1/connection/get.ts [hunk 5]
```diff
@@ -53,87 +53,73 @@ export const getConnection = asyncWrapper<GetConnection>(async (req, res) => {
...
-                        error: {
-                            code: 'unknown_provider_config',
-                            message: credentialResponse.error.message
-                        }
-                    });
-                    break;
-            }
-        }
+        const errorLog = await errorNotificationService.auth.get(connectionRes.response.id as number);
+        res.status(200).send({ errorLog, provider: integration.provider, connection: connectionRes.response });
```
Comment (TBonnin) [inline packages/server/lib/controllers/v1/connection/get.ts:112]: should we return something else than 200 in case of error?

File: packages/shared/lib/services/connection.service.ts [hunk 6]
```diff
@@ -1063,36 +1034,39 @@ class ConnectionService {
...
+                const errorWithPayload = new NangoError(error!.type, { connection: connectionWithoutCredentials });
 
                 // there was an attempt to refresh the token so clear it from the queue
                 // of connections to refresh if it failed
-                await this.updateLastFetched(connection.id);
+                await this.updateLastFetched(connection.id!);
 
+                /**
+                 * Be careful this Error contains credentials
+                 */
```
Comment (TBonnin) [inline packages/shared/lib/services/connection.service.ts]: is this comment correct? errorWithPayload should not contain credentials
```
                const { credentials, ...connectionWithoutCredentials } = copy;
                const errorWithPayload = new NangoError(error!.type, { connection: connectionWithoutCredentials });
```

File: packages/server/lib/hooks/connection/post-connection.ts (hunk 2 above)
Comment (khaliqgant) [reply packages/server/lib/hooks/connection/post-connection.ts:26]: No, should be fine to just use `getConnection` here

File: packages/shared/lib/services/connection.service.ts (hunk 4 above)
Comment (bodinsamuel) [reply packages/shared/lib/services/connection.service.ts:994]: it's modifying the credentials by reference since we don't fetch anymore inside, it's best practice to copy (even though it's a shallow copy)

File: packages/server/lib/controllers/v1/connection/get.ts (hunk 5 above)
Comment (bodinsamuel) [reply packages/server/lib/controllers/v1/connection/get.ts:112]: The thing is in the UI it's still used to display the page, before that we returned an error by did not respected that and still displayed the page. My take on this: it should be a totally different endpoint.

File: packages/shared/lib/services/connection.service.ts (hunk 6 above)
Comment (bodinsamuel) [reply packages/shared/lib/services/connection.service.ts]: oops leftover from the first fix ty

File: packages/server/lib/controllers/v1/connection/get.ts (hunk 5 above)
Comment (TBonnin) [reply packages/server/lib/controllers/v1/connection/get.ts:112]: ok. can you add a comment?

File: packages/server/lib/refreshConnections.ts (hunk 3 above)
Comment (bodinsamuel) [reply packages/server/lib/refreshConnections.ts:39]: yes, it's just eslint best practice rules when you return void https://typescript-eslint.io/rules/no-confusing-void-expression/

**Review by TBonnin — APPROVED** [review]
